[supervisor]: gitpod config changed analytics

akosyakov opened the pull request:
## Description
<!-- Describe your changes in detail -->

This PR adds analytics for .gitpod.yml changes. We are track when changes are persisted to a disk and debounce them for 5s. After last change we analyse changes to top-level elements and report an event for each. It allows us to gain insights in what users are changing on top-level, e.g `ports` or `tasks`,   and avoid a lot of traffic at the same time. Later we can do more fine grained analysis to see which object properties, like `ports.name`.

I had to resolve couple issues which were discovered during review or were blocking testing:
- enable analytics env inside workspace in preview environments to verify with segment
- don't fire change events for syntactically broken .gitpod.yml files, thank you @andreafalzetti for noticing it
- adding git hook which keeps goland types up to date with gitpod schema, otherwise we will miss new properties, thank you @mustard-mh for brining forward that yaml parser only see properties which are defined in goland

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes https://github.com/gitpod-io/gitpod/issues/6894

## How to test
<!-- Provide steps to test this PR -->

- Open staging untrusted.
- Start a workspace, open .gitpod.yml yet, wait for 5s.
- You should not yet see any `gitpod_config_changed` events.
- Now edit .gitpoy.yml, then you finish wait for 5s.
- Now you should see `gitpod_config_changed` events.
- Edit .gitpod.yml to add some syntax errors and semantic changes, then you finish wait for 5s.
- You should not yet see any `gitpod_config_changed` events.
- Edit .gitpod.yml to fix syntax errors, then you finish wait for 5s.
- Now you should see `gitpod_config_changed` events for semantic changes.

Check for regressions:
- Add new ports and check there are detected in the ports view.

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
NONE
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

## Werft options:
<!--
Optional annotations to add to the werft job.

* with-preview - whether to create a preview environment for this PR
-->
- [x] /werft with-preview
- [x] /werft analytics=segment|TEZnsG4QbLSxLfHfNieLYGF4cDwyFWoe

roboquat commented:
@akosyakov: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

akosyakov commented:
/werft run with-clean-slate-deployment=true

akosyakov commented:
/werft run

akosyakov commented:
Important if you want to rebuild then you start a Gitpod workspace and do following:
`werft run github -s .werft/jobs/build/installer/installer.ts -s .werft/jobs/build/installer/post-process.sh -f`

Because of security concerns, changes to werft scripts are not picked by default, but scripts from main are used. We need to side load them.

andreafalzetti commented on components/supervisor/pkg/config/gitpod-config-analytics.go:
```diff
@@ -0,0 +1,145 @@
+}
+
+func (a *ConfigAnalyzer) Analyse(current *gitpod.GitpodConfig) <-chan struct{} {
+	a.log.Debug("gitpod config analytics: schedulling")
```
```suggestion
	a.log.Debug("gitpod config analytics: scheduling")
```

akosyakov commented on components/supervisor/pkg/config/gitpod-config-analytics.go:
```diff
@@ -0,0 +1,145 @@
+func (a *ConfigAnalyzer) computeFields(configs ...*gitpod.GitpodConfig) []string {
+	defer func() {
+		if err := recover(); err != nil {
+			a.log.WithField("error", err).Error("failed to compute gitpod config fields")
```
`gitpod config analytics:`

akosyakov commented on components/supervisor/pkg/config/gitpod-config-analytics.go:
```diff
@@ -0,0 +1,145 @@
+func (a *ConfigAnalyzer) valueByField(config *gitpod.GitpodConfig, field string) interface{} {
+	defer func() {
+		if err := recover(); err != nil {
+			a.log.WithField("error", err).WithField("field", field).Error("failed to retrieve value from gitpod config")
```
`gitpod config analytics:`

akosyakov commented on components/supervisor/pkg/supervisor/supervisor.go:
```diff
@@ -1513,82 +1512,39 @@ func socketActivationForDocker(ctx context.Context, wg *sync.WaitGroup, term *te
+func analyseConfigChanges(ctx context.Context, wscfg *Config, w analytics.Writer, cfgobs config.ConfigInterface, gitpodAPI gitpod.APIInterface) {
+	info, err := gitpodAPI.GetWorkspace(ctx, wscfg.WorkspaceID)
+	if err != nil {
+		log.WithError(err).Error("failed to track config changes")
```
`gitpod config analytics: failed to fetch workspace owner`

andreafalzetti commented on components/supervisor/pkg/supervisor/supervisor.go:
```diff
@@ -1513,82 +1512,39 @@ func socketActivationForDocker(ctx context.Context, wg *sync.WaitGroup, term *te
-		}
-		return fmt.Sprintf("%x", h.Sum(nil)), nil
+func analyseConfigChanges(ctx context.Context, wscfg *Config, w analytics.Writer, cfgobs config.ConfigInterface, gitpodAPI gitpod.APIInterface) {
+	info, err := gitpodAPI.GetWorkspace(ctx, wscfg.WorkspaceID)
```
Is there any benefit on calling the API instead of using `os.Getenv("GITPOD_WORKSPACE_ID")`?

andreafalzetti replied:
never mind, I just noticed it's for the owner

akosyakov replied:
yes, GITPOD_WORKSPACE_ID is internal implementation, Config abstracts away from it.

mustard-mh replied:
Yes, @akosyakov always said that once we use `GITPOD_*` env var [[1](https://github.com/gitpod-io/gitpod/pull/8313#discussion_r810007160)] 😂 

But I'm not sure if we have a chance to deprecate it since many users are using Gitpod's env vars. So maybe it's safe somehow? Anyway it must be best if we don't use it

akosyakov replied:
there is somethere an issue to provide `gp info` instead and remove all `GITPOD_*` from docs

mustard-mh commented on components/supervisor/pkg/config/gitpod-config-analytics.go:
```diff
@@ -0,0 +1,145 @@
+	a.prev = current
+}
+
+func (a *ConfigAnalyzer) computeFields(configs ...*gitpod.GitpodConfig) []string {
```
This function will always returns all keys of `.gitpod.yml` since they are init-ed as a struct, i.e. pointer->nil string->"" int->0.

Or this is as expected?

akosyakov replied:
I don't think it is true, at least I don't see it while testing. I can add more tests.

mustard-mh replied:
This is from the first test case

![image](https://user-images.githubusercontent.com/20944364/181787373-c120d24e-5b0b-4c2e-b58f-e72f3d85b4e7.png)

mustard-mh commented:
We only get the diffs for the first layer of fields, is it enough? Maybe `inner-loop` can re-use this `ConfigAnalyzer` but it required more detail of diff (i.e. tasks.?.name don't need to trigger, but tasks.?.command need)

andreafalzetti commented:
When the YAML file is invalid, additional events are fired, loom: https://www.loom.com/share/e140a1decc6c4019bbb41785ec280c72

mustard-mh commented:
> When the YAML file is invalid, additional events are fired, loom: https://www.loom.com/share/e140a1decc6c4019bbb41785ec280c72

Yes, it treated this object as nil 🙈, so diff from the previous one, will return all previous fields.

akosyakov commented:
> We only get the diffs for the first layer of fields, is it enough? Maybe inner-loop can re-use this ConfigAnalyzer but it required more detail of diff (i.e. tasks.?.name don't need to trigger, but tasks.?.command need)

It is an idea to have deep structural analysis, but with another PR.

sagor999 reviewed (approved):
reviewed one file that workspace team is code owner of and it LGTM :smile:

akosyakov commented:
/hold

because it does not handle invalid files well and it turned out we have a bug generally in gitpod config watching, i.e. fire nil config for invalid

akosyakov commented on components/gitpod-protocol/go/gitpod-config-types.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2020 Gitpod GmbH. All rights reserved.
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
```
This file auto-generated. I added git hook which regenerates it on commit, see [.pre-commit-config.yaml](https://github.com/gitpod-io/gitpod/pull/11653/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9)

akosyakov commented on components/gitpod-protocol/data/gitpod-schema.json:
```diff
@@ -261,140 +261,20 @@
-                            "description": "Configure JVM options, for instance '-Xmx=4096m'."
-                        }
-                    }
+                    "$ref": "#/definitions/jetbrainsProduct",
```
Refactored to extract a common properties to the definition and stabilize code generation.

felladrin commented:
ℹ️ Reviewing it now.

felladrin commented on components/supervisor/pkg/supervisor/supervisor.go:
```diff
@@ -1513,82 +1512,39 @@ func socketActivationForDocker(ctx context.Context, wg *sync.WaitGroup, term *te
+						Identity: analytics.Identity{UserID: info.Workspace.OwnerID},
+						Event:    "gitpod_config_changed",
+						Properties: map[string]interface{}{
+							"key":         field,
```
**Observation (non-blocking)**: Opening a workspace without .gitpod.yml and then running `gp init` triggered 6 `gitpod_config_changed` at once, one for each of the keys:
- MainConfiguration
- WorkspaceLocation
- Ports
- CheckoutLocation
- Tasks
- ExperimentalNetwork

<img width="507" alt="image" src="https://user-images.githubusercontent.com/418083/183590308-b86f4f80-8cf8-4875-a690-43b6e21f1f26.png">

The content of the generated .gitpod.yml was the following:

```yml
# List the start up tasks. Learn more https://www.gitpod.io/docs/config-start-tasks/
tasks:
  - init: echo 'init script' # runs during prebuild
    command: echo 'start script'

# List the ports to expose. Learn more https://www.gitpod.io/docs/config-ports/
ports:
  - port: 3000
    onOpen: open-preview
```

So I expected it would trigger just two events keys: "Tasks" and "Ports".

akosyakov replied:
rrr, it is a bug, let me have a look a bit later, what is going on

felladrin reviewed (approved):
Working as advertised 🎉 

I left just a comment about events triggered when the workspace has no `.gitpod.yml` and we run `gp init`, which is an edge case.

akosyakov commented:
> I left just a comment about events triggered when the workspace has no .gitpod.yml and we run gp init, which is an edge case.

let me verify first that it does not happen then a user creates an empty file as well, it does not look good

akosyakov commented:
@felladrin I pushed the fix when https://werft.gitpod-dev.com/job/gitpod-build-ak-identiy-config-changed.38 is ready you can try again.

andreafalzetti reviewed (approved):
Great work @akosyakov 🙏 

It works as expected when the `.gitpod.yml` is invalid. I've noticed that when creating or deleting the whole file it does detect changes for the entire file but I'd say it's kind of expected and I don't consider it a bug.

akosyakov commented:
@loujaybee @andreafalzetti Do you think it would be useful to add an attribute like `operation` to distnguish what happened, i.e. `added`, `updated`,  `removed`? I looked right now at analytics it can be misleading, i.e. you can see 2 modifications to tasks, but final maybe was removing so actually tasks are not used. For analytics it maybe interesting to focus only on `added` and `updated`? It is rather an easy fix.

akosyakov commented:
> @loujaybee @andreafalzetti Do you think it would be useful to add an attribute like operation to distinguish what happened, i.e. added, updated, removed? I looked right now at analytics it can be misleading, i.e. you can see 2 modifications to tasks, but final maybe was removing so actually tasks are not used. For analytics it maybe interesting to focus only on added and updated? It is rather an easy fix.

Actually I'm wrong we cannot distinguish `removed` from `updated` because our goland types using string types not pointers. It will be quite impactful change, let's leave it for later.

akosyakov commented:
/unhold

akosyakov commented:
/werft run
